File: ShellBoost.Samples/ShellBoost.Samples.Overview/SimplePngItem.cs
using System;
using ShellBoost.Core;
using ShellBoost.Core.Utilities;
using ShellBoost.Core.WindowsPropertySystem;
using ShellBoost.Core.WindowsShell;

namespace ShellBoost.Samples.Overview
{
    public class SimplePngItem : ShellItem
    {
        public SimplePngItem(SimpleFolder parent, string text, ulong key)
            : base(parent, new StringKeyShellItemId(text))
        {
            // this is needed for icon
            ItemType = IOUtilities.PathGetExtension(text);

            Thumbnail = new SimpleThumbnail(this, key);
        }

        public new SimpleThumbnail Thumbnail { get => (SimpleThumbnail)base.Thumbnail; set => base.Thumbnail = value; }

        public void ClearCache()
        {
            // note this can fail with a sharing violation if the files are accessed (by the Shell itself or something else)
            Thumbnail.DeleteCache(96, 96);
            Thumbnail.DeleteCache(256, 256);
        }

        public override bool TryGetPropertyValue(PropertyKey key, out object value)
        {
            // https://docs.microsoft.com/en-us/windows/win32/properties/props-system-thumbnailcacheid
            // NOTE: cache id cannot be 0
            if (key == Core.WindowsPropertySystem.System.ThumbnailCacheId)
            {
                value = Thumbnail.Key;
                return true;
            }

            return base.TryGetPropertyValue(key, out value);
        }

        // we use a ShellBoost utility class to download (and cache) the file
        public override ShellContent GetContent() => new WebCacheShellContent(
                Parent.Root.FolderServer.Configuration.GetDefaultWebFileCache(),
                SimpleThumbnail.GetUrl(256, 256, Thumbnail.Key),
                new WebFileCacheRequestOptions { DontForceServerCheck = true }
            );
    }

    public class SimpleThumbnail : ShellThumbnail
    {
        public SimpleThumbnail(SimplePngItem item, ulong key)
        {
            if (item == null)
                throw new ArgumentNullException(nameof(item));

            Item = item;
            Key = key;
        }

        public ulong Key { get; }
        public SimplePngItem Item { get; }

        // This example uses a well-known Windows icon that represents a generic image.
        public override ShellThumbnailAsIcon GetAsIcon(GILIN inOptions) => new ShellThumbnailAsIcon("shell32.dll") { Index = 325 };

        // don't get surprised if you see cute cats
        // check the doc on https://loremflickr.com/
        public static string GetUrl(int width, int height, ulong key) => "https://loremflickr.com/" + width + "/" + height + "?lock=" + key;

        public string GetCachePath(int width, int height)
        {
            var url = GetUrl(width, height, Key);

            // we use a ShellBoost utility class to download (and cache) the file
            var cache = Item.Parent.FolderServer.Configuration.GetDefaultWebFileCache();
            return cache.Download(url, new WebFileCacheRequestOptions { DontForceServerCheck = true });
        }

        public bool DeleteCache(int width, int height)
        {
            var url = GetUrl(width, height, Key);
            var cache = Item.Parent.FolderServer.Configuration.GetDefaultWebFileCache();
            return cache.Delete(url);
        }

        public override ShellThumbnailAsImage GetAsImage(int width, int height)
        {
            var path = GetCachePath(width, height);
            return path != null ? new ShellThumbnailAsImage(path) : null;
        }
    }
}

File: ShellBoost.Samples/ShellBoost.Samples.CppOverview/SimplePngItem.cpp
#include "pch.h"
#include "SimplePngItem.h"

using namespace ShellBoost::Core::Utilities;

SimplePngItem::SimplePngItem(ShellFolder^ parent, String^ text, UInt64 key)
	:ShellItem(parent, gcnew StringKeyShellItemId(text))
{
	// this is needed for icon
	ItemType = IOUtilities::PathGetExtension(text);

	Thumbnail = gcnew SimpleThumbnail(this, key);
}

bool SimplePngItem::TryGetPropertyValue(WindowsPropertySystem::PropertyKey key, Object^% value)
{
	// https://docs.microsoft.com/en-us/windows/win32/properties/props-system-thumbnailcacheid
	// NOTE: cache id cannot be 0
	if (key == WindowsPropertySystem::System::ThumbnailCacheId)
	{
		value = Thumbnail->Key;
		return true;
	}

	return ShellItem::TryGetPropertyValue(key, value);
}

void SimplePngItem::ClearCache()
{
	// note this can fail with a sharing violation if the files are accessed (by the Shell itself or something else)
	Thumbnail->DeleteCache(96, 96);
	Thumbnail->DeleteCache(256, 256);
}

ShellContent^ SimplePngItem::GetContent()
{
	auto options = gcnew WebFileCacheRequestOptions();
	options->DontForceServerCheck = true;

	auto content = gcnew WebCacheShellContent(
		Parent->Root->FolderServer->Configuration->GetDefaultWebFileCache(),
		SimpleThumbnail::GetUrl(256, 256, Thumbnail->Key),
		options);
	return content;
}

SimpleThumbnail::SimpleThumbnail(SimplePngItem^ item, UInt64 key)
{
	if (item == nullptr)
		throw gcnew ArgumentNullException(L"item");

	Item = item;
	Key = key;
}

ShellThumbnailAsIcon^ SimpleThumbnail::GetAsIcon(GILIN inOptions)
{
	auto icon = gcnew ShellThumbnailAsIcon(L"shell32.dll");
	icon->Index = 325;
	return icon;
}

// don't get surprised if you see cute cats
// check the doc on https://loremflickr.com/
String^ SimpleThumbnail::GetUrl(int width, int height, UInt64 key)
{
	return L"https://loremflickr.com/" + width + L"/" + height + L"?lock=" + key;
}

String^ SimpleThumbnail::GetCachePath(int width, int height)
{
	auto url = GetUrl(width, height, Key);

	// we use a ShellBoost utility class to download (and cache) the file
	auto cache = Item->Parent->FolderServer->Configuration->GetDefaultWebFileCache();
	auto options = gcnew WebFileCacheRequestOptions();
	options->DontForceServerCheck = true;
	return cache->Download(url, options);
}

bool SimpleThumbnail::DeleteCache(int width, int height)
{
	auto url = GetUrl(width, height, Key);
	auto cache = Item->Parent->FolderServer->Configuration->GetDefaultWebFileCache();
	return cache->Delete(url);
}

ShellThumbnailAsImage^ SimpleThumbnail::GetAsImage(int width, int height)
{
	auto path = GetCachePath(width, height);
	return path != nullptr ? gcnew ShellThumbnailAsImage(path) : nullptr;
}


